File: llm_generated_code/ChatGPT/Python/Level2/Q7.py
class TreeNode:
    def __init__(self, value=0, next=None):
        self.value = value
        self.next = next

def has_cycle(start_node):
    slow = start_node  # Slow pointer
    fast = start_node  # Fast pointer

    while fast is not None and fast.next is not None:
        slow = slow.next         # Move slow by one step
        fast = fast.next.next    # Move fast by two steps

        if slow == fast:  # Cycle detected
            return True

    return False  # No cycle

# Example setup
# Creating nodes for testing
a = TreeNode(1)
b = TreeNode(2)
c = TreeNode(3)
d = TreeNode(4)

# Creating a cycle for testing: 1 -> 2 -> 3 -> 4 -> 2 ...
a.next = b
b.next = c
c.next = d
d.next = b  # Cycle here

# Check if there is a cycle
if has_cycle(a):
    print("The path has a cycle.")
else:
    print("The path is safe to explore.")

File: llm_generated_code/ChatGPT/CPP/Level2/Q7.cpp
#include <iostream>
using namespace std;

struct Tree {
    int id;
    Tree* next;
    Tree(int val) : id(val), next(nullptr) {}
};

bool hasCycle(Tree* head) {
    if (!head) return false;

    Tree* slow = head;
    Tree* fast = head;

    while (fast && fast->next) {
        slow = slow->next;          // Move slow by 1 step
        fast = fast->next->next;     // Move fast by 2 steps

        if (slow == fast) {          // Cycle detected
            return true;
        }
    }

    return false;  // No cycle found
}

int main() {
    // Example forest path
    Tree* root = new Tree(1);
    root->next = new Tree(2);
    root->next->next = new Tree(3);
    root->next->next->next = new Tree(4);
    root->next->next->next->next = root->next;  // Creating a cycle

    if (hasCycle(root)) {
        cout << "The path leads to a cycle. Be cautious!" << endl;
    } else {
        cout << "The path is safe to explore." << endl;
    }

    return 0;
}

